Replace `hardfork_at` with `eras` object, and add node_era

Anviking opened the pull request:
# Issue Number

ADP-676, ADP-681


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->
- [x] Instead of storing ProtocolParameters in DB, just keep them in a `TVar`
- [x] Remove problem where the wallet might use genesis-derived protocol parameters and era on wallet start up, until the node first changed tip.
- [x] Replace `hardfork_at` with info of all eras
- [x] Add new `node_era` field in GET /network/information

- [ ] <s>See if we can order the eras correctly in the `era` object (for when the JSON is viewed by humans)</s> Seems impossible to do.

# Comments

- Could add absolute `slot` field also
- Arguably a bit sad that the eras get sorted seemingly alphabetically in the json

```json
$ cardano-wallet network parameters
{
    "slot_length": {
        "quantity": 1,
        "unit": "second"
    },
    "decentralization_level": {
        "quantity": 74,
        "unit": "percent"
    },
    "genesis_block_hash": "5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb",
    "blockchain_start_time": "2017-09-23T21:44:51Z",
    "desired_pool_number": 500,
    "epoch_length": {
        "quantity": 432000,
        "unit": "slot"
    },
    "eras": {
        "shelley": {
            "epoch_start_time": "2020-07-29T21:44:51Z",
            "epoch_number": 208
        },
        "mary": null,
        "byron": {
            "epoch_start_time": "2017-09-23T21:44:51Z",
            "epoch_number": 0
        },
        "allegra": {
            "epoch_start_time": "2020-12-16T21:44:51Z",
            "epoch_number": 236
        }
    },
    "active_slot_coefficient": {
        "quantity": 5,
        "unit": "percent"
    },
    "security_parameter": {
        "quantity": 2160,
        "unit": "block"
    },
    "minimum_utxo_value": {
        "quantity": 1000000,
        "unit": "lovelace"
    }
}
```

<!-- Additional comments or screenshots to attach if any -->

<!--
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Jira will detect and link to this PR once created, but you can also link this PR in the description of the corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
 ✓ Finally, in the PR description delete any empty sections and all text commented in <!--, so that this text does not appear in merge commit messages.
-->

Anviking committed:
Remove ProtocolParameters from DB

Anviking committed:
Replace hardfork_at with eras

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -2190,6 +2190,11 @@ taggedSumTypeOptions base opts = base
 
+explicitNothingRecordTypeOptions :: Aeson.Options
+explicitNothingRecordTypeOptions = defaultRecordTypeOptions
+    { omitNothingFields = False
```
Not sure why this is enabled by default…

"Is the `mary` key not present because the wallet is not supporting `mary`, or because the `mary` fork has not yet occurred?"

Explicit `null` for the latter case seems much clearer.

KtorZ replied:
I agree that the absence of a field doesn't have the same meaning as a `null` value. I think that in the context of our API indeed, we most likely always want fields corresponding to Maybe values to be present :thinking:

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Network.hs:
```diff
@@ -51,7 +51,8 @@ spec = describe "SHELLEY_NETWORK" $ do
             , expectField #minimumUtxoValue (`shouldBe` Quantity minUTxOValue)
-            , expectField #hardforkAt (`shouldNotBe` Nothing)
+            , expectField (#eras . #shelley) (`shouldNotBe` Nothing)
+            , expectField (#eras . #byron) (`shouldNotBe` Nothing)
```
:+1:

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Network.hs:
```diff
@@ -51,7 +51,8 @@ spec = describe "SHELLEY_NETWORK" $ do
             , expectField #minimumUtxoValue (`shouldBe` Quantity minUTxOValue)
-            , expectField #hardforkAt (`shouldNotBe` Nothing)
+            , expectField (#eras . #shelley) (`shouldNotBe` Nothing)
+            , expectField (#eras . #byron) (`shouldNotBe` Nothing)
```
Integration tests can in practice be run on any shelley-based era. So simply testing that _at least_ `byron` and `shelley` are not null, should be pretty good.

paweljakubas replied:
but then you will not receive information about mary and allegra, right? :thinking:

Anviking replied:
Ok, fine, we can probably make the expectation dependent on the `LOCAL_CLUSTER_ERA` value, by storing it in the test `Context`

Anviking replied:
Hm, seems unexpectedly tedious though… (Can't use a shelley type in cardano-wallet-core-integration) Will look more at it later.

Anviking commented on specifications/api/swagger.yaml:
```diff
@@ -1410,6 +1410,30 @@ components:
+        mary: *nullableEpochInfo
+      description: |
+
+        If and when each era started or will start.
+
+        The object is keyed by era names. The values either describe the epoch boundary
+        when the era starts (can be in the future or in the past), or are null if not yet
+        confirmed on-chain.
```
Because of
> (can be in the future or in the past)

wouldn't it make sense to provide a `era` field either here, or in `/network/information`?

A non-null `mary` does not necessarily mean that we're in Mary. We could also be just before Mary.

paweljakubas replied:
I thought it would go like this : we know something like mary will happen (null), then we know when it happens (by specifying era it starts), and we also now current slot/era. So everything is fine;-)

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -749,15 +751,30 @@ data ApiNetworkParameters = ApiNetworkParameters
     , desiredPoolNumber :: !Word16
     , minimumUtxoValue :: !(Quantity "lovelace" Natural)
-    , hardforkAt :: !(Maybe ApiEpochInfo)
+    , eras :: !ApiEraInfo
```
why not to have 
```
eras: [ApiEraInfo]
...
data Eras = Byron | Shelley | Allegra | Mary deriving (Eq, Ord, Show)

data ApiEraInfo = ApiEraInfo {
   hardForkToEra :: !Era
   hardFromEra :: !(Maybe Era)
   timeOfTransition:: !(Maybe ApiEpochInfo)
}
```
?

paweljakubas replied:
I think it is more robust in a sense that when new hard forks arrive they will just add new element to list. But maybe there is some rationale behind this way of storing era transitions?

Anviking replied:
@KtorZ advocated for an object keyed by era-name since Daedalus wants to do stuff like `isAllegra` and `isShelley`, and already have to be highly aware of e.g. the ordering of them.

Although with https://github.com/input-output-hk/cardano-wallet/pull/2468#discussion_r565335344 in mind, I wonder if this changes 🤔

Anviking replied:
Seems we should indeed add some kind of  `currentEra` field somewhere.

Maybe we should make this a list then, yes 🤔 

Off the top of my head I would probably just do
```
data ApiEraInfo = ApiEraInfo {
   eraName :: !Era
   start :: !(Maybe ApiEpochInfo)
}
```
since `hardFromEra` doesn't provide any extra information.

Anviking replied:
In-doubt I would stick to the current object format. I think it still has its qualities too. If you feel strongly about it, feel free to disagree.

But I imagine, if a need arises where a list would be great for Daedalus, we can change it then. Currently the need for this is unclear.

paweljakubas replied:
you got me wrong Johannes. I like your proposal, just was thinking loudly and sharing some ideas ;-) no pun intended.great job!

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -953,6 +955,38 @@ newtype ActiveSlotCoefficient
+-- It might be cumbersome to work with this type. /But/ we don't need to. A
+-- product of @Maybe@ is both what we can query from the node, and
+-- what we need to provide in the wallet API.
+data EraInfo info = EraInfo
```
If you opt for my suggestion of ApiEraInfo then this part should also comply

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -709,19 +709,20 @@ mkTipSyncClient tr np localTxSubmissionQ onPParamsUpdate onInterpreterUpdate = d
-                _ ->
-                    LSQry $ QueryAnytimeMary GetEraStart
+            eraBounds <- W.EraInfo
+                <$> LSQry (QueryAnytimeByron GetEraStart)
```
:+1:

paweljakubas commented on specifications/api/swagger.yaml:
```diff
@@ -1410,6 +1410,30 @@ components:
+      <<: *epochInfo
+      nullable: True
+
+    ApiEraInfo: &ApiEraInfo
```
the same here in case ApiEraInfo is refactored

KtorZ commented on lib/core/test/data/Cardano/Wallet/Api/ApiEraInfo.json:
```diff
@@ -0,0 +1,161 @@
+                "epoch_start_time": "1891-03-26T08:00:00Z",
+                "epoch_number": 53
+            },
+            "byron": null,
```
That would be problematic in practice :)

KtorZ reviewed (approved):
Sounds good, provided that we make sure of the behavior on start-up with regards to protocol parameters. I'd expect the server to block on a particular request until protocol parameters become available.

Anviking committed:
To-be-safe: block until we get recent PParams

We never want to use the old genesis parameters just because the wallet
just started.

The onTipUpdate will be run immediately on startup, and will fetch
both interpreter, and protocol parameters.

To be safe, and for readability, let's replace the TVar+initial val with
an empty TMVar though.

Anviking committed:
Ensure the node era is immediately known

Anviking committed:
Add node_era field to GET /network/information

Anviking committed:
Workaround problem from marking ApiEraInfo fields required
